### PR TITLE
Fix bug where CEL listOfString.join() results in unexpected error

### LINKED ISSUE #117590: CEL: listOfStrings.join() fails due to internal conversion error
### What happened?

Added this test to `staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go`:

```go
		{name: "string lists",
			obj:    objs([]interface{}{"a", "b", "c"}),
			schema: schemas(listType(&stringType)),
			valid: []string{
				"self.val1.join('-') == 'a-b-c'"
			},
		},
```

this resulted in:

```
Invalid value: "object": internal error: interface conversion: interface {} is []interface {}, not []string evaluating rule: self.val1.join('-') == 'a-b-c'
        --- FAIL: TestValidationExpressions/string_lists/self.val1.join('-')_==_'a-b-c' (0.00s)
```

### What did you expect to happen?

The  test should pass.

### How can we reproduce it (as minimally and precisely as possible)?

Add the above test

### Anything else we need to know?

_No response_

### Kubernetes version

All versions where CEL is used.

### Cloud provider

All

### OS version

Doesn't matter

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Without string.join() it can be difficult to leverage the `messageExpression` field added to CRD validation rules and ValidatingAdmissionPolicy in 1.27.

#### Which issue(s) this PR fixes:

Fixes #117590

#### Special notes for your reviewer:

We may be able to simplify this in the future depending on the outcome of https://github.com/google/cel-go/issues/688.  For now, this introduces an isolated workaround.

#### Does this PR introduce a user-facing change?

```release-note
Fix bug where `listOfStrings.join()` in CEL expressions resulted in an unexpected internal error.
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go [hunk 1]
```diff
@@ -339,6 +339,17 @@ func TestValidationExpressions(t *testing.T) {
 				"self.val1 + [4, 5] == [1, 2, 3, 4, 5]",
 			},
 		},
+		{name: "string lists",
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:342]: do we need a negative test that `obj: objs([]interface{}{"a", 1, "c"}),` errors when join is attempted?

Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:342]: The CEL compiler prevents mixed type lists via https://github.com/google/cel-go/blob/d39523c445fcf548ef6529b63c1a1045cc373d37/cel/options.go#L129, which we have enabled.

I could test that non-string lists are not eligible to be joined, but it seems excessive to test that all functions are only available on the receivers that they are defined for.

Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:342]: ok, if we guard against it and test for that guard, that's fine.

**Comment by jpbetz** [comment]
> /lgtm Might worth to backport?

https://github.com/kubernetes/kubernetes/pull/117596 backports to 1.27, which is where this is needed most, since it helps a lot for `messageExpression`.  If anyone thinks this should go back further let me know.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go (hunk 1 above)
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go:346]: SG. I've added the tests for the guard (for both lists and maps).

File: staging/src/k8s.io/apiserver/pkg/cel/common/values.go [hunk 2]
```diff
@@ -425,7 +426,22 @@ var _ = traits.Lister(&unstructuredList{})
...
-		return t.elements, nil
+		switch t.itemsSchema.Type() {
+		// Workaround for https://github.com/kubernetes/kubernetes/issues/117590 until we
+		// resolve the desired behavior in cel-go via https://github.com/google/cel-go/issues/688
+		case "string":
+			var result []string
+			for _, e := range t.elements {
+				s, ok := e.(string)
+				if !ok {
+					return nil, fmt.Errorf("unexpected all elements to be of type string, but got %T", e)
```
Comment (alvaroaleman) [inline staging/src/k8s.io/apiserver/pkg/cel/common/values.go:437]:
```suggestion
					return nil, fmt.Errorf("expected all elements to be of type string, but got %T", e)
```

**Comment by jpbetz** [comment]
/assign @deads2k @lavalamp 

Would either of you be wiling to approve?

**Pull request merged**